agent/eventstream/eventstream.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2014-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package handler deals with appropriately reacting to all ACS messages as well
+// as maintaining the connection to ACS.
+package eventstream
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/cihub/seelog"
+	"golang.org/x/net/context"
+)
+
+type eventHandler func(...interface{}) error
+
+// EventStream waiting for events and notifying the listeners by invoking
+// the handler that listeners registered
+type EventStream struct {
+	name         string
+	open         bool
+	event        chan interface{}
+	handlers     map[string]eventHandler
+	ctx          context.Context
+	handlersLock sync.RWMutex
+	statusLock   sync.RWMutex
+}
+
+func NewEventStream(name string, ctx context.Context) *EventStream {
+	return &EventStream{
+		name:     name,
+		event:    make(chan interface{}, 1),
+		ctx:      ctx,
+		open:     false,
+		handlers: make(map[string]eventHandler),
+	}
+}
+
+// Subscribe adds the handler to be called into EventStream
+func (eventStream *EventStream) Subscribe(name string, handler eventHandler) error {
+	eventStream.handlersLock.Lock()
+	defer eventStream.handlersLock.Unlock()
+
+	if _, ok := eventStream.handlers[name]; ok {
+		return fmt.Errorf("handler name has already existed, %v", name)
+	}
+
+	eventStream.handlers[name] = handler
+	return nil
+}
+
+// broadcast calls all handler's handler function
+func (eventStream *EventStream) broadcast(event interface{}) {
+	eventStream.handlersLock.RLock()
+	defer eventStream.handlersLock.RUnlock()
+
+	seelog.Debugf("Event stream %v received events, broadcasting to listeners...", eventStream.name)
```

Comment:
(multiple occurrences) Can you change `%v` to `%s` since these are just strings?